crypto/stark-verifier-ethereum/contracts/stark_verifier.sol

```diff
@@ -0,0 +1,193 @@
+            bytes32 oods_point,
+            bytes32[] memory oods_coefficients,
+            bytes32[] memory eval_points
+        ) = write_data_and_read_random(proof, constraint_parameters, coin);
+        // Preform the proof of work check
+        require(check_proof_of_work(coin, proof.pow_nonce, constraint_parameters.pow_bits), 'POW Failed');
+        // Read the query indecies from the coin
+        uint8 eval_domain_log_size = constraint_parameters.log_trace_length + constraint_parameters.log_blowup;
+        uint64[] memory queries = get_queries(coin, eval_domain_log_size, constraint_parameters.number_of_queries);
+    }
+
+    // This function write to the channel and reads from the channel to get the randomized data
+    function write_data_and_read_random(
+        StarkProof memory proof,
+        ConstraintParameters memory constraint_parameters,
+        PublicCoin.Coin memory coin
+    ) internal pure returns (bytes32[] memory, bytes32, bytes32[] memory, bytes32[] memory) {
+        // Write the trace root to the coin
+        coin.write_bytes32(proof.trace_root);
+        // Read random constraint coefficentrs from the coin
+        bytes32[] memory constraint_coeffiencents = coin.read_many_bytes32(
+            2 * constraint_parameters.number_of_constraints
+        );
+        // Write the evaluated constraint root to the coin
+        coin.write_bytes32(proof.constraint_root);
+        // Read the oods point from the coin
+        bytes32 oods_point = coin.read_bytes32();
+        // Write the trace oods values to the coin
+        coin.write_many_bytes32(proof.trace_oods_values);
+        // Write the constraint oods values to the coin
+        coin.write_many_bytes32(proof.constraint_oods_values);
+        // Read the oods coeffients from the random coin
+        bytes32[] memory oods_coefficients = coin.read_many_bytes32(
+            proof.trace_oods_values.length + proof.constraint_oods_values.length
+        );
+
+        // Writes the fri merkle roots and reads eval points from the coin
+        bytes32[] memory eval_points = new bytes32[](constraint_parameters.fri_layout.length);
+        for (uint256 i; i < constraint_parameters.fri_layout.length; i++) {
+            coin.write_bytes32(proof.fri_roots[i]);
+            eval_points[i] = coin.read_bytes32();
+        }
+        // Write the claimed last layer points a set of coeffient for the final layer fri check
+        // NOTE - This is a fri layer so we have to write the whole thing at once
+        coin.write_bytes(abi.encodePacked(proof.last_layer_coeffiencts));
+        return (constraint_coeffiencents, oods_point, oods_coefficients, eval_points);
+    }
+
+    // Given a coin and a nonce hashes the random form the coin and checks that the proof of works passes
+    // NOTE - This function also advances the coin by writing the pow_nonce to it
+    function check_proof_of_work(PublicCoin.Coin memory coin, bytes8 pow_nonce, uint8 pow_bits)
```

Comment:
Proof of work is a separate sub-protocol. It's cleaner to have it in a separate library so it can be independently tested.